magma_access_gateway_installer/agw_network_configurator.py

```diff
@@ -0,0 +1,234 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+import os
+from subprocess import check_call, check_output
+from typing import Optional
+
+import ipcalc  # type: ignore[import]
+import yaml
+from systemd.journal import JournalHandler  # type: ignore[import]
+
+logger = logging.getLogger(__name__)
+logger.addHandler(JournalHandler())
+logger.setLevel(logging.INFO)
+
+
+class AGWInstallerNetworkConfigurator:
+
+    DNS_ADDRESSES = "8.8.8.8 208.67.222.222"
+    BOOT_GRUB_GRUB_CFG_PATH = "/boot/grub/grub.cfg"
+    CLOUD_INIT_YAML_PATH = "/etc/netplan/50-cloud-init.yaml"
+    ETC_DEFAULT_GRUB_PATH = "/etc/default/grub"
+    ETC_SYSTEMD_RESOLVED_CONF_PATH = "/etc/systemd/resolved.conf"
+    INTERFACES_DIR = "/etc/network/interfaces.d"
+
+    def __init__(
+        self,
+        network_interfaces: list,
+        eth0_address: Optional[str],
+        eth0_gw_ip_address: Optional[str],
+    ):
+        self.eth0_address = eth0_address
+        self.eth0_gw_ip_address = eth0_gw_ip_address
+        self.network_interfaces = network_interfaces
+        self.reboot_needed = False
+
+    def configure_network_interfaces(self):
+        """Configures network interfaces if necessary."""
+        if not self._network_interfaces_are_named_eth0_and_eth1:
+            self._update_interfaces_names_in_cloud_init()
+            self._configure_grub()
+            self._update_grub_cfg()
+        if not self._dns_configured:
+            self._configure_dns()
+        if not self._network_interfaces_config_files_exist:
+            self._create_interfaces_config_files()
+        if self._netplan_installed:
+            self._remove_netplan()
+        if not self._networking_service_enabled:
+            self._enable_networking_service()
+
+    @property
+    def _network_interfaces_are_named_eth0_and_eth1(self) -> bool:
+        """Checks whether available network interfaces are named correctly."""
+        return all(interface in self.network_interfaces for interface in ["eth0", "eth1"])
+
+    def _update_interfaces_names_in_cloud_init(self):
+        """Changes names of network interfaces in /etc/netplan/50-cloud-init.yaml to ethX."""
+        with open(self.CLOUD_INIT_YAML_PATH, "r") as cloud_init_file:
+            cloud_init_content = yaml.safe_load(cloud_init_file)
```

Comment:
I was thinking maybe we can  this variable something like `initial_cloud_init_content`, then take a copy of it and call that one `modified_cloud_init_content` and play around with this one. I'm saying this because now `cloud_init_content` evolves from being one thing to another.

Reply:
Done.